### PR TITLE
add nubOrd and friends

### LINKED ISSUE #439: add ordNub somewhere in containers
After dropping the standard ordNub (i.e. this: https://github.com/nh2/haskell-ordnub) code into a codebase for the zillionth time I went and checked the history of why it never made it into a standard lib. It can't go into base because it requires containers or reimplementing `Set` just for this purpose, which is silly. It doesn't necessarily belong in the `Data.Set` module. But perhaps it could fall into a new module, like `Data.SetUtils` or `Data.ContainerUtils` or the like? Anyway, this package is the most natural place for it to live, even if it doesn't quite fit, and everyone ends up using it everywhere, so it would be really nice to just have a standard for it...

### PR DESCRIPTION
**Pull request by gbaz**
resolves #439

I think I have the asymptotics for `nubInt` right, but a sanity check wouldn't hurt...

### CODE REVIEW

**Comment by treeowl** [comment]
Could you add QuickCheck properties?

File: Data/Containers/ListUtils.hs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+-- In particular, it keeps only the first occurrence of each element. By using a 'Set' internally
+-- it has better asymptotics than the standard 'nub' function.
+nubOrd :: (Ord a) => [a] -> [a]
+nubOrd l = go Set.empty l
+  where
+    go _ [] = []
+    go s (x:xs) = if x `Set.member` s then go s xs
+                                      else x : go (Set.insert x s) xs
+
+-- | The `nubOrdOn` function behaves just like `nubOrd` except it preforms comparisons not on the
```
Comment (treeowl) [inline Data/Containers/ListUtils.hs]: Performs, not preforms.

File: Data/Containers/ListUtils.hs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+-- it has better asymptotics than the standard 'nub' function.
+nubInt :: [Int] -> [Int]
+nubInt l = go IntSet.empty l
+  where
+    go _ [] = []
+    go s (x:xs) = if x `IntSet.member` s then go s xs
+                                         else x : go (IntSet.insert x s) xs
+
+-- | The `nubIntOn` function behaves just like 'nubInt' except it preforms comparisons not on the
+-- original datatype, but a user-specified projection from that datatype to 'Int'.
```
Comment (treeowl) [inline Data/Containers/ListUtils.hs]: Performs, not preforms.

File: Data/Containers/ListUtils.hs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+nubIntOn :: (a -> Int) -> [a] -> [a]
+nubIntOn f l = go IntSet.empty l
```
Comment (treeowl) [inline Data/Containers/ListUtils.hs]: Eta reduce. And it's probably slightly faster to make `f` an argument to `go`, here and above.

**Comment by treeowl** [comment]
Actually, I changed my mind. We want to let `f` have the chance to inline, so keep it pulled out.

**Comment by treeowl** [comment]
However however, inlining `f` is only useful if we make this participate in list fusion. So let's do that too.

**Comment by gbaz** [comment]
Added tests, fixed typo, eta reduced.

I did not add fusion. I took a run at it and it wasn't straightforward. Doing a foldr doesn't really work because it reverses the direction in which things are dropped. So that means making it a good consumer is hard. Further, the way in which we end up building out of the fold (which involves taking a snd of the result, since we need to pass around a pair) is rather ugly and I think gets in the way of making it a good producer.

**Comment by treeowl** [comment]
You can definitely make it a good consumer nicely.

nubOrd xs = foldr go stop xs S.empty

**Comment by gbaz** [comment]
I don't know what that means. In any case, I tried, and didn't get it working. If you know how to get it working, I suggest we merge first, and then you make a further patch.

**Comment by treeowl** [comment]
Sure.

On Jan 26, 2018 1:46 AM, "gbaz" <notifications@github.com> wrote:

I don't know what that means. In any case, I tried, and didn't get it
working. If you know how to get it working, I suggest we merge first, and
then you make a further patch.

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/haskell/containers/pull/515#issuecomment-360697951>,
or mute
the thread

**Comment by gbaz** [comment]
Ok, is this good to merge then?

**Comment by treeowl** [comment]
Thanks.

**Pull request merged**